### PR TITLE
fix: Change emoji navigation with arrow keys

### PR DESCRIPTION
**Pull request by mananjadhav**
### Details

- Allow the user to go to the first emoji on click of the right arrow
- Allow the user to go to the search input clicking on the left arrow while first emoji is selected.

### Fixed Issues

$ https://github.com/Expensify/App/issues/7035

### Tests
1. Tested `ArrowRight` to ensure that it selects the first emoji when the cursor is at the end of the string
2. `ArrowDown` should ensure that it goes to the first emoji irrespective of the cursor position
3. `ArrowUp` and `ArrowLeft` would select the text in the input and rest of the cursor works fine.

- [x] Verify that no errors appear in the JS console

### QA Steps
1. Open any chat
2. Open the emoji picker and type a search term
3. If the cursor is at the end of the input, and click the `right arrow` key
4. It should select the first emoji
5. When the first emoji is selected, and you click `left arrow`, it should focus the search input and select the text

- [x] Verify that no errors appear in the JS console

### Tested On

- [x] Web
- [x] Mobile Web
- [ ] Desktop
- [ ] iOS
- [ ] Android

#### Web

https://user-images.githubusercontent.com/3069065/149575845-efb0bb4b-148f-4b3a-a5b6-e031c62dda49.mov

#### Mobile Web

NA

#### Desktop

https://user-images.githubusercontent.com/3069065/149575877-56a5625c-0b75-4756-899f-9eae40e87856.mov

#### iOS

NA

#### Android

NA

### CODE REVIEW

**Comment by rushatgabhane** [comment]
Yeah sure, go ahead

**Comment by mananjadhav** [comment]
The `selection` prop isn't working fine with `TextInputFocussable`. It automatically resets to `{start: 0, end: 0}` even if I am setting it `{start: 5, end: 5}`. 

I need this resolved because we want the emoji to be highlighted only if the cursor is at the end of the string. Once this is resolved rest is done.

**Comment by rushatgabhane** [comment]
> Allow the user to go to the first emoji on click of the right arrow

Just to clarify, we only wanna go to the first emoji when cursor is at the end of search query.

**Comment by rushatgabhane** [comment]
> The selection prop isn't working fine with TextInputFocussable. It automatically resets to {start: 0, end: 0} even if I am setting it {start: 5, end: 5}.

I think we should skip it then. React Native has an open issue on controlled selection https://github.com/facebook/react-native/issues/29063

Let's do it the way up arrow is done. What do you think?

**Comment by mananjadhav** [comment]
> When search is refocused, we should highlight the search query, so it can be replaced by typing.

That is what is handled with ArrowUp that the whole string is selected. This can be done.

**Comment by mananjadhav** [comment]
> That is what is handled with ArrowUp that the whole string is selected. This can be done.

@rushatgabhane Looks like I was wrong. This isn't how `ArrowUp` and `ArrowDown` work. Just confirmed on staging.

1. Irrespective of where my cursor is in the search string, and I click on `ArrowDown`, it'll go to the first emoji.
2. Clicking on `ArrowUp` on the first emoji, takes the cursor to the beginning of the search string

While the second one is easy, I feel we might need a `selection` prop to know the cursor position so that either it goes to the next character or to the first emoji.

**Comment by rushatgabhane** [comment]
Gothu, but to verify.

- On `ArrowRight`, go to first emoji if and only if cursor is at end.
- On `ArrowLeft` from first emoji, highlight the text.
- Change `ArrowUp` from just focus to highlight the text.

> selection prop for the first one to know the cursor position so that either it goes to the next character or to the first emoji. Any other suggestions to handle this?

Umm, I think you can check if cursor is at end using `onSelectionChange`.

**Comment by mananjadhav** [comment]
> * On `ArrowRight`, go to first emoji if and only if cursor is at end.
> * On `ArrowLeft` from first emoji, highlight the text.
> * Change `ArrowUp` from just focus to highlight the text.
> 

@rushatgabhane  Yeah the updated PR covers the above cases.

File: src/pages/home/report/EmojiPickerMenu/index.js [hunk 1]
```diff
@@ -178,7 +204,16 @@ class EmojiPickerMenu extends Component {
 
         // Arrow Down enable arrow navigation when search is focused
```
Comment (luacmartins) [inline src/pages/home/report/EmojiPickerMenu/index.js:205]: We should update the comment to include Arrow right

**Review by luacmartins — COMMENTED** [review]
Just left a small comment. Otherwise LGTM and tests well. Great work @mananjadhav!

File: src/pages/home/report/EmojiPickerMenu/index.js [hunk 2]
```diff
@@ -178,7 +204,16 @@ class EmojiPickerMenu extends Component {
...
         if (this.searchInput && this.searchInput.isFocused() && this.state.filteredEmojis.length) {
-            if (arrowKey !== 'ArrowDown') {
+            if (['ArrowDown', 'ArrowRight'].indexOf(arrowKey) === -1) {
```
Comment (rushatgabhane) [inline src/pages/home/report/EmojiPickerMenu/index.js:207]: Can this be changed to something like `arrowKey !== 'ArrowDown' || arrowKey !== 'ArrowRight'`

File: src/pages/home/report/EmojiPickerMenu/index.js [hunk 3]
```diff
@@ -169,6 +184,17 @@ class EmojiPickerMenu extends Component {
...
     }
 
+    focusInputWithTextSelect() {
+        if (!this.searchInput) {
+            return;
+        }
+
+        this.setState({selectTextOnFocus: true});
+
+        // Reaching start of the list, arrow left/up set the focus to searchInput.
```
Comment (rushatgabhane) [inline src/pages/home/report/EmojiPickerMenu/index.js:194]: This comment should be removed. Instead add JSDoc for this function.

**Review by rushatgabhane — CHANGES_REQUESTED** [review]
Nice work! Minor suggestions

**Review by rushatgabhane — APPROVED** [review]
LGTM! 🎉

cc: @luacmartins

**Review by luacmartins — APPROVED** [review]

**Comment by luacmartins** [comment]
Thanks! @mananjadhav and @parasharrajat!

**Pull request merged**